crashtracker/src/collector/crash_handler.rs

```diff
@@ -611,29 +598,33 @@ fn handle_posix_signal_impl(
 ///     will not yet be stored.  This would lead to unexpected behaviour for the
 ///     user.  This should only matter if something crashes concurrently with
 ///     this function executing.
-pub fn register_crash_handlers() -> anyhow::Result<()> {
+pub fn register_crash_handlers(config: &CrashtrackerConfiguration) -> anyhow::Result<()> {
     if !OLD_HANDLERS.load(SeqCst).is_null() {
         return Ok(());
     }
 
-    let config_ptr = CONFIG.load(SeqCst);
-    anyhow::ensure!(!config_ptr.is_null(), "No crashtracking config");
-    let (config, _config_str) = unsafe { config_ptr.as_ref().context("config ptr")? };
-
-    unsafe {
-        if config.create_alt_stack {
-            create_alt_stack()?;
-        }
-        let sigbus = register_signal_handler(signal::SIGBUS, config)?;
-        let sigsegv = register_signal_handler(signal::SIGSEGV, config)?;
-        let boxed_ptr = Box::into_raw(Box::new(OldHandlers { sigbus, sigsegv }));
-
-        let res = OLD_HANDLERS.compare_exchange(ptr::null_mut(), boxed_ptr, SeqCst, SeqCst);
+    if config.create_alt_stack {
+        unsafe { create_alt_stack()? };
+    }
+    let mut old_handlers = HashMap::new();
+    // TODO: if this fails, we end in a situation where handlers have been registered with no chain
+    for signum in &config.signals {
+        let signal = signal_from_signum(*signum)?;
         anyhow::ensure!(
-            res.is_ok(),
-            "TOCTTOU error in crashtracker::register_crash_handlers"
+            !old_handlers.contains_key(signum),
+            "Handler already registered for {signal}"
         );
+        let handler = unsafe { register_signal_handler(signal, config)? };
+        old_handlers.insert(*signum, (signal, handler));
     }
+    let boxed_ptr = Box::into_raw(Box::new(old_handlers));
+
+    let res = OLD_HANDLERS.compare_exchange(ptr::null_mut(), boxed_ptr, SeqCst, SeqCst);
+    anyhow::ensure!(
```

Comment:
This leaks the old_handlers, if we care?

Reply:
I don't think so--small one-time leak, right?